src/api/focus.js

```diff
@@ -111,114 +111,112 @@ class Focus {
+    return await this.checkSerialDevice(
       focusDeviceDescriptor,
       focusDeviceDescriptor.usb.bootloader
     );
   }
 
-  async waitForDFUBootloader(focusDeviceDescriptor) {
-    const delay = (ms) => new Promise((res) => setTimeout(res, ms));
+  async checkNonSerialBootloader(focusDeviceDescriptor) {
     const bootloader = focusDeviceDescriptor.usb.bootloader;
 
-    for (let attempt = 0; attempt < 10; attempt++) {
-      const deviceList = await ipcRenderer.invoke(
-        "usb.scan-for-devices",
-        bootloader.productId,
-        bootloader.vendorId
-      );
-
-      for (const device of deviceList) {
-        const pid = device.deviceDescriptor.idProduct,
-          vid = device.deviceDescriptor.idVendor;
-
-        if (pid == bootloader.productId && vid == bootloader.vendorId) {
-          logger("focus").info("bootloader found", {
-            device: bootloader,
-            function: "waitForDFUBootloader",
-          });
-          return true;
-        }
-      }
+    const deviceList = await ipcRenderer.invoke(
+      "usb.scan-for-devices",
+      bootloader.productId,
+      bootloader.vendorId
+    );
 
-      logger("focus").debug("bootloader not found, waiting 2s", {
-        function: "waitForDFUBootloader",
-      });
-      await delay(2000);
+    for (const device of deviceList) {
+      const pid = device.deviceDescriptor.idProduct,
+        vid = device.deviceDescriptor.idVendor;
+
+      if (pid == bootloader.productId && vid == bootloader.vendorId) {
+        const newPort = Object.assign({}, device);
+        newPort.focusDeviceDescriptor = focusDeviceDescriptor;
+        newPort.focusDeviceDescriptor.bootloader = true;
+
+        logger("focus").info("bootloader found", {
+          device: bootloader,
+          function: "checkNonSerialBootloader",
+        });
+        return newPort;
+      }
     }
 
-    logger("focus").warn("bootloader not found after 10 attempts", {
-      function: "waitForDFUBootloader",
+    logger("focus").debug("bootloader not found", {
+      function: "checkNonSerialBootloader",
+      device: bootloader,
     });
-    return false;
+
+    return null;
   }
 
-  async waitForBootloader(focusDeviceDescriptor) {
+  async checkBootloader(focusDeviceDescriptor) {
     if (!focusDeviceDescriptor.usb.bootloader) {
-      this.logger.warn("No bootloader defined in the device descriptor", {
+      this.logger().warn("No bootloader defined in the device descriptor", {
         descriptor: focusDeviceDescriptor,
       });
-      return null;
+      return false;
     }
-    logger("focus").info("waiting for bootloader", {
+    logger("focus").info("checking bootloader presence", {
       descriptor: focusDeviceDescriptor,
     });
 
-    if (focusDeviceDescriptor.usb.bootloader.type == "dfu") {
-      return await this.waitForDFUBootloader(focusDeviceDescriptor);
+    if (focusDeviceDescriptor.usb.bootloader.type == "non-serial") {
```

Comment:
maybe it should just be !== "serial" rather than having a constant that's "not serial"?

Reply:
Can do. Should I lift it out into a constant, too, or are we ok with a string for now?

(note: this part of the code I will be reworking next, so this whole bootloader type thing may either be completely gone, or look very different in a week)